homeassistant/components/onvif/camera.py

```diff
@@ -372,6 +376,57 @@ def __init__(self, hass, config):
         except exceptions.ONVIFError as err:
             _LOGGER.error("Couldn't setup camera '%s'. Error: %s", self._name, err)
 
+    async def async_obtain_snapshot_uri(self):
+        """Set the snapshot uri for the camera."""
+        _LOGGER.debug(
+            "Connecting with ONVIF Camera: %s on port %s", self._host, self._port
+        )
+
+        try:
+            _LOGGER.debug("Retrieving profiles")
+
+            media_service = self._camera.create_media_service()
+
+            profiles = await media_service.GetProfiles()
+
+            _LOGGER.debug("Retrieved '%d' profiles", len(profiles))
+
+            if self._profile_index >= len(profiles):
+                _LOGGER.warning(
+                    "ONVIF Camera '%s' doesn't provide profile %d."
+                    " Using the last profile.",
```

Comment:
Thinking about this (after merge I know...)
This can become a problem. E.g., when using ONVIF from a DVR, the different profile will actually provide different camera's. Using the last profile as a fallback can cause a totally different image to be shown, right?

Reply:
I copied this code from get_input_uri function.
I don't understand exactly what you mean.

But there is probably something wrong.
If the profile selected is out of the profiles array range, self._profile_index becomes -1.
What happens on return line?
`return profiles[self._profile_index].token`

Reply:
-1 means last item in the array.

Reply:
Hmm yeah, we actually do the same for getting the input URL. So I guess the behavior would be the same.

Although I believe that that behavior might cause issues, a simple search shows that it never has been reported as such. Probably because most use cases are with a single camera (all profiles deliver the same image on different qualities) and people with DVR's probably know which profiles to supply.

So 👍 let's keep it like so I would say.

Reply:
The only test I can do is set up a profile that doesn't exist and it works.
```
2020-03-24 22:04:08 DEBUG (MainThread) [custom_components.onvif.camera] Setting up the ONVIF camera platform
2020-03-24 22:04:08 DEBUG (MainThread) [custom_components.onvif.camera] Constructing the ONVIFHassCamera
2020-03-24 22:04:08 DEBUG (MainThread) [custom_components.onvif.camera] Importing dependencies
2020-03-24 22:04:08 DEBUG (MainThread) [custom_components.onvif.camera] Setting up the ONVIF camera component
2020-03-24 22:04:08 DEBUG (MainThread) [custom_components.onvif.camera] Setting up the ONVIF camera device @ '192.168.0.200:80'
2020-03-24 22:04:08 DEBUG (MainThread) [custom_components.onvif.camera] Updating service addresses
2020-03-24 22:04:11 DEBUG (MainThread) [custom_components.onvif.camera] Setting up the ONVIF device management service
2020-03-24 22:04:12 DEBUG (MainThread) [custom_components.onvif.camera] Retrieving current camera date/time
2020-03-24 22:04:12 DEBUG (MainThread) [custom_components.onvif.camera] TimeZone for date/time: UTC
2020-03-24 22:04:12 DEBUG (MainThread) [custom_components.onvif.camera] Camera date/time: 0120-03-24 21:04:12+00:00
2020-03-24 22:04:12 DEBUG (MainThread) [custom_components.onvif.camera] Camera date/time in UTC: 0120-03-24 21:04:12+00:00
2020-03-24 22:04:12 DEBUG (MainThread) [custom_components.onvif.camera] System date/time: 2020-03-24 21:04:12.811324+00:00
2020-03-24 22:04:12 DEBUG (MainThread) [custom_components.onvif.camera] Connecting with ONVIF Camera: 192.168.0.200 on port 80
2020-03-24 22:04:12 DEBUG (MainThread) [custom_components.onvif.camera] Retrieving profiles
2020-03-24 22:04:14 DEBUG (MainThread) [custom_components.onvif.camera] Retrieved '1' profiles
2020-03-24 22:04:14 WARNING (MainThread) [custom_components.onvif.camera] ONVIF Camera 'Kitchen' doesn't provide profile 3. Using the last profile.
2020-03-24 22:04:14 DEBUG (MainThread) [custom_components.onvif.camera] Using profile index '-1'
2020-03-24 22:04:14 DEBUG (MainThread) [custom_components.onvif.camera] Retrieving stream uri
2020-03-24 22:04:15 DEBUG (MainThread) [custom_components.onvif.camera] ONVIF Camera Using the following URL for CAM cucina: rtsp://<user>:<password>@192.168.0.200/ch0_0.h264
2020-03-24 22:04:15 DEBUG (MainThread) [custom_components.onvif.camera] Connecting with ONVIF Camera: 192.168.0.200 on port 80
2020-03-24 22:04:15 DEBUG (MainThread) [custom_components.onvif.camera] Retrieving profiles
2020-03-24 22:04:17 DEBUG (MainThread) [custom_components.onvif.camera] Retrieved '1' profiles
2020-03-24 22:04:17 DEBUG (MainThread) [custom_components.onvif.camera] Using profile index '-1'
2020-03-24 22:04:17 DEBUG (MainThread) [custom_components.onvif.camera] Retrieving snapshot uri
2020-03-24 22:04:19 DEBUG (MainThread) [custom_components.onvif.camera] ONVIF Camera Using the following URL for CAM cucina snapshot: http://<user>:<password>@192.168.0.200:8080/cgi-bin/snapshot.sh?res=high
2020-03-24 22:04:19 DEBUG (MainThread) [custom_components.onvif.camera] Setting up the ONVIF PTZ service
2020-03-24 22:04:19 DEBUG (MainThread) [custom_components.onvif.camera] PTZ is not available
2020-03-24 22:04:19 DEBUG (MainThread) [custom_components.onvif.camera] Completed set up of the ONVIF camera component
2020-03-24 22:04:19 DEBUG (MainThread) [custom_components.onvif.camera] Camera 'Kitchen' added to hass
```